File: 00383_ransom/solution.go
func canConstruct(ransomNote string, magazine string) bool {
    hm :=  make(map[rune]int)
    for _, d := range magazine {
        if count, ok := hm[d]; ok {
            hm[d] = count + 1
        } else {
            hm[d] = 0
        }
    }

    for _, d := range ransomNote {
        count, ok := hm[d]
        if ! ok || count == -1 {
            return false
        } else {
            hm[d] = count - 1
        }
    }

    return true
}

File: 00383_ransom/solution.rs
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        // hashmap to hold counts of chars in magazine
        use std::collections::HashMap;
        
        let mut h = HashMap::new();
        
        for c in magazine.chars() {
            let c_count = h.entry(c).or_insert(0);
            *c_count += 1;
        }
        
        // now pass through ransom_note
        for c in ransom_note.chars() {
            let c_count = h.entry(c).and_modify(|x| *x -= 1).or_insert(-1);
            if *c_count == -1 {
                return false;
            }
        }
        return true;
    }
}
